kernel: Add a "nocache" read-write memory section

aurel32 opened the pull request:
Add a "nocache" read-write memory section that is configured to
not be cached. This memory section can be used to perform DMA
transfers when cache coherence issues are not optimal or can not
be solved using cache maintenance operations.

This is currently only supported on ARM Cortex M7 with MPU.

Fixes #2927

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

aurel32 committed:
kernel: Add a "nocache" read-write memory section

Add a "nocache" read-write memory section that is configured to
not be cached. This memory section can be used to perform DMA
transfers when cache coherence issues are not optimal or can not
be solved using cache maintenance operations.

This is currently only supported on ARM Cortex M7 with MPU.

Fixes #2927

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

ioannisg commented on arch/arm/core/cortex_m/mpu/arm_mpu_v7_internal.h:
```diff
@@ -106,6 +106,17 @@ static inline void _get_mpu_ram_region_attr(arm_mpu_region_attr_t *p_attr,
 	p_attr->rasr = _get_region_attr(1, ap, 1, 1, 1, 1, 0, size);
 }
 
+static inline void _get_mpu_nocache_region_attr(arm_mpu_region_attr_t *p_attr,
```
This is still SRAM, so _ram_ should be included in the function name, IMO

ioannisg commented on include/arch/arm/cortex_m/mpu/arm_core_mpu_dev.h:
```diff
@@ -32,6 +32,9 @@ extern "C" {
  */
 enum {
+#ifdef CONFIG_NOCACHE_MEMORY
+	NOCACHE_MEMORY_DATA_REGION,
```
Is this going to be a "kernel" section only?

aurel32 replied:
I believe so, at least it's one I have done for the MPU permissions, and it matches the original case of doing DMA within the drivers.

Now this is an RFC because I believe we want something that can be available on other MPU or even other architectures. If there are some uses case that would also need to setup uncached regions in the user memory, we can still change that.

andrewboie commented:
Still thinking about this, have you considered adding cache-ability to the flags that can be set for memory domain partitions instead of making this a special case?

aurel32 commented:
> Still  thinking about this, have you considered adding cache-ability to the  flags that can be set for memory domain partitions instead of making  this a special case?

I haven't considered that, and that might indeed be a good idea. I don't really know about memory domain partitions, so just gave a quick look. It seems that currently memory domains is tight to userspace, It is a limitation that can be lifted? Also is the memory domain always active in privileged mode? The primary use case I see for a nocache region is for drivers.

pizi-nordic commented on arch/arm/core/cortex_m/mpu/arm_mpu_v7_internal.h:
```diff
@@ -106,6 +106,17 @@ static inline void _get_mpu_ram_region_attr(arm_mpu_region_attr_t *p_attr,
+	 */
+	(void) base;
+
+	p_attr->rasr = _get_region_attr(1, ap, 0, 0, 0, 1, 0, size);
```
These attributes implies strongly ordered access to the region. Is that needed for regular memory?
IMHO normal, non-cacheable region (TEX=001, C=0, B=0) will give better performance (but you will have to use barriers to ensure that all memory access completed).

aurel32 replied:
I agree the attributes should be relaxed. Your suggestion looks good to me, I'll update that in the next version.

aurel32 commented on include/arch/arm/cortex_m/scripts/linker.ld:
```diff
@@ -286,6 +286,21 @@ SECTIONS
+		*(.nocache)
+		*(".nocache.*")
+		. = ALIGN(32);
+		. = ALIGN( 1 << LOG2CEIL(__nocache_end - __nocache_start));
```
Yes, this is correct, but this is actually required by the ARM MPU which only supports region aligned to their size.

qianfan-Zhao replied:
I had a discuss with @easonxiang 
The align is used only when CONFIG_MPU enabled. Seems we can disable align if MPU is not enabled to save memorys.

aurel32 replied:
There is no other way to disable the cache for a given region than using the MPU. That's why NOCACHE_MEMORY depends on enabling the MPU:

```
config NOCACHE_MEMORY
       bool "Support for uncached memory"
       depends on ARM_MPU && CPU_CORTEX_M7

```

qianfan-Zhao replied:
got it. this is the difference between nocache and ramfunc sections.

ioannisg reviewed (changes requested):
I am wondering if we should follow the generic approach of defining top-arch, arch-agnostic Kconfig symbols (like NOCACHE_MEMORY), but then arch-specific "intermediate" ones.

Otherwise, it looks weird that kernel/Kconfig would contain an option that depends on ARM_MPU and CORTEX_M7

aurel32 commented:
I have just pushed a new version, which I hope addresses the comment from this PR, but also from PR #11228.

I haven't introduced support for ARMv8-M as the MPU documentation is not clear about how to set cacheability attributes and also says:

> Many features of the ARMv8-M architecture are not implemented in ARMv8-M implementations, for example, Caches, Write Buffers and Tightly Coupled Memories (TCMs) are architectural features that are not currently in any ARMv8-M implementations.

ioannisg commented on arch/Kconfig:
```diff
@@ -26,6 +26,7 @@ config ARM
 	bool "ARM architecture"
 	select ARCH_HAS_THREAD_ABORT
 	select HAS_DTS
+	select ARCH_HAS_NOCACHE_MEMORY_SUPPORT if ARM_MPU && CPU_CORTEX_M7
```
Why is this selection needed, if the symbol is selected in CPU_CORTEX_M7?

aurel32 replied:
That's a leftover from previous version, I'll fix it in the next version.

ioannisg commented on arch/arm/core/cortex_m/mpu/arm_mpu_v7_internal.h:
```diff
@@ -106,6 +106,17 @@ static inline void _get_mpu_ram_region_attr(arm_mpu_region_attr_t *p_attr,
+	p_attr->rasr = _get_region_attr(1, ap, 1, 0, 0, 1, 0, size);
+}
+
 /**
```
Ideally, all remaining code should not be compiled in if just CONFIG_NOCHACE is enabled

ioannisg reviewed (changes requested):
Added a few comments - looks generally OK

aurel32 commented:
Thanks for the review. I have just pushed a new version that should address the two issues.

ioannisg commented:
> That's true, what you are doing here is more towards boot-time memory region configuration that depends on data whose bounds are only known after link-time, in which case memory domains wouldn't be appropriate as they are per-thread and tightly coupled to user mode.

I agree that we, perhaps need, here a boot-time configuration of a fixed memory region.
However, it is not a bad idea to have memory domains with partitions with cache-related flags.

ioannisg commented on arch/arm/core/cortex_m/Kconfig:
```diff
@@ -70,6 +70,7 @@ config CPU_CORTEX_M7
 	# Omit prompt to signify "hidden" option
 	select ARMV7_M_ARMV8_M_MAINLINE
 	select ARMV7_M_ARMV8_M_FP if CPU_HAS_FPU
+	select ARCH_HAS_NOCACHE_MEMORY_SUPPORT if ARM_MPU
```
Could you move this "ARCH_HAS..." option together with the reset of ARCH_HAS options in CPU_CORTEX_M?

ioannisg reviewed (changes requested):
Some more suggestions from me.
I still have to look at the linker script changes

aurel32 commented:
> Some more suggestions from me.

I have just updated the PR with your suggestions.

ioannisg commented on arch/arm/core/cortex_m/mpu/arm_mpu_v7_internal.h:
```diff
@@ -233,6 +253,6 @@ static inline int _mpu_buffer_validate(void *addr, size_t size, int write)
 #endif /* CONFIG_USERSPACE */
 
-#endif /* USERSPACE || MPU_STACK_GUARD || APPLICATION_MEMORY */
+#endif /* USERSPACE || MPU_STACK_GUARD || APPLICATION_MEMORY || NOCACHE_MEMORY */
```
Isn't this more than 80 characters?

ioannisg commented on arch/arm/core/cortex_m/mpu/arm_mpu_v7_internal.h:
```diff
@@ -105,6 +107,24 @@ static inline void _get_mpu_ram_region_attr(arm_mpu_region_attr_t *p_attr,
 }
+#endif /* USERSPACE || MPU_STACK_GUARD || APPLICATION_MEMORY */
+
+#ifdef CONFIG_NOCACHE_MEMORY
```
minor: please use #if defined(CONFIG_NOCACHE_MEMORY) and #endif /* CONFIG_NO_CACHE_MEMORY */

ioannisg commented on include/linker/ram-nocache.ld:
```diff
@@ -0,0 +1,21 @@
+#ifdef CONFIG_MPU_REQUIRES_POWER_OF_TWO_ALIGNMENT
+	. = ALIGN( 1 << LOG2CEIL(_nocache_ram_end - _nocache_ram_start));
+#endif
+#ifdef CPU_HAS_ARM_MPU
```
It is not, really, CPU_HAS_ARM_MPU, rather, ARM_MPU, because HW capabilities do not imply that we build Zephyr with ARM MPU support.

ioannisg replied:
**Update:**

So: if CPU_HAS_ARM_MPU and ARM_MPU are not defined, then the CONFIG_NOCACHE_MEMORY option is never defined. So, does it make sense to query these variables inside this script, if the script is only included if CONFIG_NOCACHE_MEMORY is enabled?

aurel32 replied:
The idea with this linker script was to make it generic, even if ARM is the only supported architecture for now. I can also move it to an ARM specific directory, and it can be moved back to the top include directory once support for more architectures is added.

ioannisg replied:
I believe it is better to move it to arm subdirectory; it is safer and less confusing.
We should not have ARM specific options in a generic linker script overlay.

Still - I believe you can remove the CPU_HAS_ARM_MPU - since this is a prerequisite for the inclusion of this overlay.

ioannisg commented on include/arch/arm/cortex_m/scripts/linker.ld:
```diff
@@ -283,6 +283,10 @@ SECTIONS
 #endif
 
+#if defined(CONFIG_NOCACHE_MEMORY)
+#include <linker/ram-nocache.ld>
```
I wonder: since you provide a solution for ARM only, should this file be hosted under ./arm/....?

aurel32 replied:
I provide a solution for ARM only for now, but the linker file can work for other architectures provided the default alignment is enough.

ioannisg commented on include/arch/arm/cortex_m/scripts/linker.ld:
```diff
@@ -283,6 +283,10 @@ SECTIONS
 
+#if defined(CONFIG_NOCACHE_MEMORY)
+#include <linker/ram-nocache.ld>
+#endif
```
Another thing: this looks like Kernel RAM with no-cache attribution. Should, we , then add this section after APP SMEM and APP MEMORY? Just to group all kernel SRAM together.

ioannisg commented on include/linker/section_tags.h:
```diff
@@ -25,6 +25,10 @@
 
+#if defined(CONFIG_NOCACHE_MEMORY)
+#define __nocache __in_section_unique(_NOCACHE_SECTION_NAME)
+#endif /* CONFIG_ARM */
```
This needs a fix.

ioannisg reviewed (changes requested):
Reviewed the linker scripts, added few suggestions.
Also: #11528  attempts to standardize section aligning for MPU in ARM, so this PR should be re-based if #11528 gets merged, first.

ioannisg commented on include/linker/ram-nocache.ld:
```diff
@@ -0,0 +1,21 @@
+SECTION_PROLOGUE(_NOCACHE_SECTION_NAME,(NOLOAD),)
+{
+#ifdef CONFIG_MPU_REQUIRES_POWER_OF_TWO_ALIGNMENT
+	. = ALIGN( 1 << LOG2CEIL(_nocache_ram_end - _nocache_ram_start));
```
Could you use _nocache_ram_size here? and below?

ioannisg commented on arch/arm/core/Kconfig:
```diff
@@ -22,6 +22,7 @@ config CPU_CORTEX_M
 	select ARCH_HAS_TRUSTED_EXECUTION if ARM_TRUSTZONE_M
 	select ARCH_HAS_STACK_PROTECTION if ARM_MPU || CPU_CORTEX_M_HAS_SPLIM
 	select ARCH_HAS_USERSPACE if ARM_MPU
+	select ARCH_HAS_NOCACHE_MEMORY_SUPPORT if ARM_MPU && CPU_CORTEX_M7
```
One more thing. ARM_MPU covers also the NXP-variant. But no implementation is provided for nxp-mpu.

aurel32 replied:
Good point. I do not have a way to test this on an NPX MPU, so I guess for now I'll just limit it to the ARM MPU.

ioannisg replied:
Then I suppose you should fix the Kconfig dependency, too reflect this.

aurel32 commented:
@ioannisg , I have just pushed a new version that should address your comment. Anyway we should wait for PR #11528 to be merged first to rebase it.

ioannisg commented on include/arch/arm/cortex_m/scripts/linker.ld:
```diff
@@ -371,6 +371,17 @@ SECTIONS
 	_app_smem_rom_start = LOADADDR(_APP_SMEM_SECTION_NAME);
 #endif  /* CONFIG_APP_SHARED_MEM */
 
+	/* Non-cached region of RAM */
```
Please, add this section after the application memory, and before the shared memory section. The reason has been discussed in #11528: we would like all the static (fixed) memory regions to be in the beginning of SRAM (e.g. APP memory, no-cache, section, execute-from-SRAM section), leaving a continuous memory area on which we can define dynamic regions (e.g. thread stack).

aurel32 replied:
Ok. I have just moved it.
